src/merkle/tiered_smt/storage.rs

```diff
@@ -0,0 +1,94 @@
+use super::{BTreeMap, ContentType, Leaf, NodeIndex, Vec, Word};
+use core::{cell::RefCell, fmt};
+
+#[derive(Debug, Default)]
+pub struct Storage {
+    leaves: RefCell<BTreeMap<NodeIndex, Vec<Leaf>>>,
+    nodes: RefCell<BTreeMap<NodeIndex, Word>>,
+    types: RefCell<BTreeMap<NodeIndex, ContentType>>,
+}
```

Comment:
1. Done! However, I think it's important we keep the function as fallible so we design the advice interfaces accordingly. wdyt?
2. That's true. We do have a minimal overhead if we use vectors as `[foo].to_vec()` for single elements as this will exist only in the stack, but then it's still an overhead. An alternative is to have an enum, but then adding an extra byte to all leaves based on an exception (that might not ever happen btw) doesn't seem right. I'm updating the code to have two maps: `tier_leaves: BTreeMap<NodeIndex, Leaf>, ordered_lists: BTreeMap<NodeIndex, Vec<Leaf>>`
3. For memory backed implementation there is indeed no strong reason to have a dedicated map for the types. I think we can benefit to keep all interfaces uniform, but we can treat that as implementation detail of the storage.

Reply:
wdyt of the interfaces like this?

```rust
    pub fn append_leaf_to_list(&mut self, leaf: Leaf, index: u64) -> Result<(), StorageError> {
        todo!()
    }

    pub fn insert_node(&mut self, index: NodeIndex, node: Word) -> Result<(), StorageError> {
        todo!()
    }

    pub fn insert_tier_leaf(&mut self, leaf: Leaf, index: NodeIndex) -> Result<(), StorageError> {
        todo!()
    }

    pub fn get_list_leaf(&self, index: u64, key: &Word) -> Result<Option<Leaf>, StorageError> {
        todo!()
    }

    pub fn get_node(&self, index: &NodeIndex) -> Result<Option<Word>, StorageError> {
        todo!()
    }

    pub fn get_tier_leaf(&self, key: &Word) -> Result<Option<(NodeIndex, Leaf)>, StorageError> {
        todo!()
    }

    pub fn get_type(&self, index: &NodeIndex) -> Result<NodeType, StorageError> {
        todo!()
    }

    pub fn get_leaf_from_list(&self, index: u64, key: &Word) -> Result<Option<Leaf>, StorageError> {
        todo!()
    }

    pub fn get_leaves_list(&self, index: &NodeIndex) -> Result<Option<Vec<Leaf>>, StorageError> {
        Ok(self.leaves.get(index).cloned())
    }
```

Reply:
I've thought of this for a while, and at this point I'm not sure what's the right interface. I could see quite a few approaches and they all require different types of interface (e.g., page-based approached described in https://github.com/0xPolygonMiden/crypto/pull/45#discussion_r1090699154 or materializing the first tier in memory as mentioned in https://github.com/0xPolygonMiden/crypto/pull/45#issuecomment-1410061948). There is also a question of how to ensure that updates are atomic in case of DB-backed storage, and probably a few other things I haven't thought of.

This makes me think that maybe we are trying to define a general interface prematurely. So, maybe for now we don't try to generalize. Let's assume that this is an in-memory variant of the Tiered SMT and implement accordingly (and keep it as simple as possible. Once we have this working (and integrated with Miden assembly implementation of SMT), we can thing about other storage options.

One potential way to describe the simplest in-memory storage is this:
```Rust
pub struct Storage {
    nodes: BTreeMap<NodeIndex, Word>,
    upper_leaves: BTreeMap<NodeIndex, Leaf>,
    lower_leaves: BTreeMap<u64, Vec<Leaf>>,
}
```
I haven't spent too much time thinking about this specific approach though - so, there might be better ways.

Reply:
A change in that direction was performed. I added a couple more auxiliary maps to keep all operations efficient, even if they are in-memory.